taskfiles/utils/cpp-lint.yaml

```diff
@@ -4,36 +4,315 @@ set: ["u", "pipefail"]
+  # @param {string} [GIT_REF="origin/main"] A `git` reference that HEAD is diff'd with.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
+  # @param {string} [OUTPUT_DIR="{{.ROOT_DIR}}/lint-clang-tidy"]
+  clang-tidy-diff:
+    internal: true
+    label: "{{.TASK}}:{{.OUTPUT_DIR}}"
+    vars:
+      GIT_REF: >-
+        {{default "origin/main" .GIT_REF}}
+      DIFF_SOURCE_FILE_PATHS_:
+        sh: >-
+          git diff --name-only --ignore-submodules "{{.GIT_REF}}" **/*.{cpp,h,hpp}
+          | xargs --no-run-if-empty realpath
+      DIFF_SOURCE_FILE_PATHS:
+        ref: >-
+          compact (splitList "\n" .DIFF_SOURCE_FILE_PATHS_)
+      FLAGS:
+        ref: "default (list) .FLAGS"
+      OUTPUT_DIR: >-
+        {{default "{{.ROOT_DIR}}/lint-clang-tidy" .OUTPUT_DIR}}
+    requires:
+      vars: ["VENV_DIR"]
+    dir: "{{.ROOT_DIR}}"
+    cmds:
+      - task: "clang-tidy-output"
+        vars:
+          FLAGS:
+            ref: ".FLAGS"
+          OUTPUT_DIR: "{{.OUTPUT_DIR}}"
+          SOURCE_FILE_PATHS:
+            ref: ".DIFF_SOURCE_FILE_PATHS"
+          VENV_DIR: "{{.VENV_DIR}}"
+
+  # Runs `clang-tidy` on files matching the given filter criteria.
+  #
+  # General parameters
+  # @param {string} VENV_DIR Python virtual environment where `clang-tidy` is installed.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
+  # @param {string} [OUTPUT_DIR="{{.ROOT_DIR}}/lint-clang-tidy"]
+  #
+  # Pattern parameters
+  # @param {string[]} ROOT_PATHS Paths recursively searched for files to run on.
+  # @param {string[]} [INCLUDE_PATTERNS] Wildcard patterns where any matching file paths (relative
+  # to `ROOT_PATHS`) will be included. By default, all paths match.
+  # @param {string[]} [EXCLUDE_PATTERNS] Wildcard patterns where any matching paths (relative to
+  # `ROOT_PATHS`) will be excluded. The search will not descend into directories matching a pattern.
+  # By default, no paths match.
+  # @param {string[]} [INCLUDE_FILENAME_PATTERNS=["*.cpp", "*.h", "*.hpp"]] Wildcard patterns where
+  # only files whose name (not the full path) matches a pattern will be included. This is useful for
+  # specifying the extensions of files to run `clang-tidy` on.
+  clang-tidy-find:
+    internal: true
+    label: "{{.TASK}}:{{.OUTPUT_DIR}}"
+    vars:
+      # General parameters
+      FLAGS:
+        ref: "default (list) .FLAGS"
+      OUTPUT_DIR: >-
+        {{default "{{.ROOT_DIR}}/lint-clang-tidy" .OUTPUT_DIR}}
+
+      # Filtering parameters
+      EXCLUDE_PATTERNS:
+        ref: "default (list) .EXCLUDE_PATTERNS"
+      INCLUDE_FILENAME_PATTERNS:
+        ref: >-
+          default (list "*.cpp" "*.h" "*.hpp") .INCLUDE_FILENAME_PATTERNS
+      INCLUDE_PATTERNS:
+        ref: "default (list) .INCLUDE_PATTERNS"
+
+      # Task variables
+      SOURCE_FILE_PATHS_:
+        # Do not quote `ROOT_PATHS` entries so that wildcards can be evaluated by the shell.
```

Comment:
```suggestion
        # NOTE:
        # - We don't quote `ROOT_PATHS` entries so that wildcards can be evaluated by the shell.
        # - `find`'s `-print0` option adds a null after the last entry, so care must be taken
        #   when splitting the output using a null as a delimiter.
```